kvserver: make MVCC GC less disruptive to foreground traffic

aayushshah15 opened the pull request:
This commit changes GC requests to no longer declare exclusive latches
at their BatchRequest's timestamp. This was already incorrect as
explained in https://github.com/cockroachdb/cockroach/issues/55293.

>The first use is broken because we acquire write latches at the batch
header's timestamp, which is set to time.Now(), so we're only
serializing with reads in the future and all other writes [1]. So we're
disruptive to everyone except who we want to serialize with – reads in
the past!

This commit makes GC requests only declare a non-mvcc exclusive latch
over the `RangeGCThresholdKey`. This is correct because:
```

// 1. We define "correctness" to be the property that a reader reading at /
// around the GC threshold will either see the correct results or receive an
// error.
// 2. Readers perform their command evaluation over a stable snapshot of the
// storage engine. This means that the reader will not see the effects of a
// subsequent GC run as long as it created a Pebble iterator before the GC
// request.
// 3. A reader checks the in-memory GC threshold of a Replica after it has
// created this snapshot (i.e. after a Pebble iterator has been created).
// 4. If the in-memory GC threshold is above the timestamp of the read, the
// reader receives an error. Otherwise, the reader is guaranteed to see a
// state of the storage engine that hasn't been affected by the GC request [5].
// 5. GC requests bump the in-memory GC threshold of a Replica as a pre-apply
// side effect. This means that if a reader checks the in-memory GC threshold
// after it has created a Pebble iterator, it is impossible for the iterator
// to point to a storage engine state that has been affected by the GC
// request.

```

As a result, GC requests should now be much less disruptive to
foreground traffic since they're no longer redundantly declaring
exclusive latches over global keys.

Resolves https://github.com/cockroachdb/cockroach/issues/55293

Release note(performance improvement): MVCC garbage collection should
now be much less disruptive to foreground traffic than before.

aayushshah15 committed:
kvserver: make MVCC GC less disruptive to foreground traffic

This commit changes GC requests to no longer declare exclusive latches
at their BatchRequest's timestamp. This was already incorrect as
explained in https://github.com/cockroachdb/cockroach/issues/55293.

>The first use is broken because we acquire write latches at the batch
header's timestamp, which is set to time.Now(), so we're only
serializing with reads in the future and all other writes [1]. So we're
disruptive to everyone except who we want to serialize with – reads in
the past!

This commit makes GC requests only declare a non-mvcc exclusive latch
over the `RangeGCThresholdKey`. This is correct because:
```

// 1. We define "correctness" to be the property that a reader reading at /
// around the GC threshold will either see the correct results or receive an
// error.
// 2. Readers perform their command evaluation over a stable snapshot of the
// storage engine. This means that the reader will not see the effects of a
// subsequent GC run as long as it created a Pebble iterator before the GC
// request.
// 3. A reader checks the in-memory GC threshold of a Replica after it has
// created this snapshot (i.e. after a Pebble iterator has been created).
// 4. If the in-memory GC threshold is above the timestamp of the read, the
// reader receives an error. Otherwise, the reader is guaranteed to see a
// state of the storage engine that hasn't been affected by the GC request [5].
// 5. GC requests bump the in-memory GC threshold of a Replica as a pre-apply
// side effect. This means that if a reader checks the in-memory GC threshold
// after it has created a Pebble iterator, it is impossible for the iterator
// to point to a storage engine state that has been affected by the GC
// request.

```

As a result, GC requests should now be much less disruptive to
foreground traffic since they're no longer redundantly declaring
exclusive latches over global keys.

Release note (performance improvement): MVCC garbage collection should
now be much less disruptive to foreground traffic than before.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/83213)
<!-- Reviewable:end -->

aayushshah15 commented:
****Next steps for this PR****

1. Benchmark YCSB-A with a low GC TTL with and without this patch
2. Confirm that the stats updates for writes and GC requests commute, and write a test for it

aayushshah15 commented:
> Benchmark YCSB-A with a low GC TTL with and without this patch

I ran ycsb-a with a 1 second GC TTL for 25 iterations (1m ramp and 5m duration) and see the following.
```
> benchstat ../benchlogs/yscbA_before ../benchlogs/yscbA_after

name  old ops/sec  new ops/sec  delta
.      51.5k ± 6%   51.3k ± 6%   ~     (p=0.658 n=25+25)

name  old p50      new p50      delta
.       1.60 ± 0%    1.60 ± 0%   ~     (all equal)

name  old p95      new p95      delta
.       4.50 ± 0%    4.50 ± 0%   ~     (all equal)

name  old p99      new p99      delta
.       41.0 ±13%    41.7 ±10%   ~     (p=0.326 n=25+25)
```

On taking a closer look, I think the results stem from the fact that there just weren't a ton of GC requests being issued in this workload (something on the order of once every 10 seconds across the cluster, will take screenshots next time). Perhaps we need a workload that writes bigger datums?

nvanbenschoten commented:
> Perhaps we need a workload that writes bigger datums?

This sounds like a good idea. It should be easy to make this change.

You could also drop down `mvccGCQueueTimerDuration` to 0s to run the queue more aggressively.

I also wonder if even with the 1-second GC TTL, the MVCC GC queue is still too unreactive for us to see the effects of this change. We never directly enqueue into it, so for a given range, it will only run at most once every scanner interval.

aayushshah15 committed:
kvserver: add microbenchmark demonstrating latchless MVCC GC improvement

This commit adds a couple simple microbenchmarks that demonstrate the
improvement in the level of  foreground traffic disruption caused by MVCC GC
after https://github.com/cockroachdb/cockroach/pull/83213.

Results:
```
> benchdiff --old=beforeLatchlessGC --new=latchlessMVCCGC --post-checkout='make buildshort' ./pkg/kv/kvserver --run=MVCCGCWithForegroundTraffic --count=20;

  pkg=1/1 iter=10/10 cockroachdb/cockroach/pkg/kv/kvserver -

name                                                     old time/op    new time/op    delta
MVCCGCWithForegroundTraffic/gc_with_reads-24               1.67ms ±18%    0.02ms ± 6%  -98.74%  (p=0.000 n=10+10)
MVCCGCWithForegroundTraffic/gc_with_reads_and_writes-24    1.10ms ±19%    0.57ms ± 4%  -48.45%  (p=0.000 n=10+10)
MVCCGCWithForegroundTraffic/noop_gc_with_reads-24          12.3µs ± 2%     7.3µs ± 1%  -40.33%  (p=0.000 n=10+9)

name                                                     old alloc/op   new alloc/op   delta
MVCCGCWithForegroundTraffic/gc_with_reads-24               17.3kB ±11%     1.7kB ± 4%  -90.44%  (p=0.000 n=10+10)
MVCCGCWithForegroundTraffic/gc_with_reads_and_writes-24    37.5kB ± 5%    24.4kB ± 3%  -34.92%  (p=0.000 n=10+10)
MVCCGCWithForegroundTraffic/noop_gc_with_reads-24          1.26kB ± 0%    1.25kB ± 1%   -0.43%  (p=0.001 n=10+10)

name                                                     old allocs/op  new allocs/op  delta
MVCCGCWithForegroundTraffic/gc_with_reads-24                  111 ± 1%        32 ± 0%  -71.08%  (p=0.000 n=9+10)
MVCCGCWithForegroundTraffic/gc_with_reads_and_writes-24       190 ± 7%       128 ± 2%  -32.37%  (p=0.000 n=10+9)
MVCCGCWithForegroundTraffic/noop_gc_with_reads-24            15.0 ± 0%      15.0 ± 0%     ~     (all equal)
```

Release note: None

aayushshah15 commented:
@aliher1911 and @erikgrinaker: I'm trying to understand the implications of https://github.com/cockroachdb/cockroach/issues/84576 over this patch. I'm missing a ton of context here but I'm trying to understand this part from Oleg's issue:
```
we can optimize range key GC requests to only get read latches to preserve range boundaries for the 
same of MVCC stats calculations, but don't prevent reads.
``` 

Is there anything you could point me to that'll help me understand why we need this additional bit of serialization, after https://github.com/cockroachdb/cockroach/pull/78980 and https://github.com/cockroachdb/cockroach/pull/76312?

aayushshah15 commented:
@nvanbenschoten: I've added a few microbenchmarks based on our discussions. Are you interested in any other scenario that you'd suggest we add in this patch?

erikgrinaker commented:
> Is there anything you could point me to that'll help me understand why we need this additional bit of serialization, after #78980 and #76312?

When we're GCing MVCC range tombstones, there are essentially two cases:

1. We're GCing individual point/range tombstones. Your work makes the point variant here latchless. However, for MVCC range tombstones, we need to serialize with all writes that overlap the MVCC range tombstone, as well as the immediate left/right neighboring keys. This is because a range key write has non-local effects, i.e. it can fragment or merge other range keys at other timestamps and at its boundaries, and this has a non-commutative effect on MVCC stats.

2. We're GCing an MVCC range tombstone and all the garbage below it by dropping a Pebble range tombstone across its key span (not yet implemented). In this case, we also need to serialize with all overlapping writes, since we otherwise could accidentally delete a newer key that was written in the meanwhile.

The crucial point here is that we only need to serialize with writers (and with other GC processes), because our writes will be invisible to MVCC readers anyway. We do not need to serialize with readers above the GC threshold. Currently, however, we serialize with both readers and writers.

I don't believe we currently have a way to take out an MVCC read/write latch that serializes with readers _below_ a timestamp, only above a timestamp, nor do we have a write-only latch. We could either build one of these, or we could possibly hack around it by taking out a read/write latch at timestamp `hlc.MaxTimestamp` which would not serialize with any readers (since there cannot be any readers above that timestamp).

aayushshah15 commented:
Thanks for that explanation, @erikgrinaker. That all makes sense.

@nvanbenschoten and @arulajmani: this is ready for a look now.

nvb reviewed (commented):
> We're GCing individual point/range keys. Your work makes the point variant here latchless. However, for MVCC range tombstones, we need to serialize with all writes that overlap the MVCC range tombstone, as well as the immediate left/right neighboring keys. This is because a range key write has non-local effects, i.e. it can fragment or merge other range keys at other timestamps and at its boundaries, and this has a non-commutative effect on MVCC stats -- if someone writes a new range key while we're GCing one below, the stats would come out wrong.

Does this PR currently address this case? 

@erikgrinaker You're saying that we need to serialize new MVCC range tombstone writes at a timestamp ts2 and the GC of old MVCC point keys at a timestamp ts1, where ts1 < ts2? Or is this only the case for the GC of old MVCC ranged keys?

> Reviewed 3 of 3 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83213#-:-N80LbN93Cm5GxwavcwN:b-59l0da)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15 and @arulajmani)

___
*[`pkg/kv/kvserver/batcheval/cmd_gc.go` line 39 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/83213#-N80HT29E9dlzZ2GZqI3:-N80HT2ADnU532o7fef9:b-gjneor) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb739440c59ef9b7ca739e1f36f7afe104e99a67/pkg/kv/kvserver/batcheval/cmd_gc.go#L39)):*
> ```Go
> ) {
> 	gcr := req.(*roachpb.GCRequest)
> 	// Extend the range key latches by feather to ensure MVCC stats computations
> ```

This is new as of https://github.com/cockroachdb/cockroach/commit/285aead75f77740df357a370aff37e91e579ff8a. Do we still need these latches or does the same reasoning that we're using to remove point latches apply to range keys as well? Maybe we should include @aliher1911 as a reviewer here.

EDIT: I see that Erik has discussed this in the thread above. We should work in some of that into the comment here.
___
*[`pkg/kv/kvserver/spanset/spanset.go` line 114 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/83213#-N80INw_CySP8zxzh9yu:-N80INw_CySP8zxzh9yv:b-ruo84b) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb739440c59ef9b7ca739e1f36f7afe104e99a67/pkg/kv/kvserver/spanset/spanset.go#L114)):*
> ```Go
> 			s.spans[sa][ss] = recycle
> 		}
> 	}
> ```

Do we need to reset `allowUndeclared` here?
___
*[`pkg/kv/kvserver/spanset/spanset.go` line 403 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/83213#-N80IjVJ9cQmkruwObK9:-N80IjVJ9cQmkruwObKA:benup7y) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb739440c59ef9b7ca739e1f36f7afe104e99a67/pkg/kv/kvserver/spanset/spanset.go#L403)):*
> ```Go
> // undeclared access to spans. This is generally set by requests that rely on
> // other forms of synchronization for correctness (e.g. GCRequest).
> func (s *SpanSet) DisableUndeclaredAccessAssertions() {
> ```

Can we use this to get rid of `DisableReaderAssertions`?
___
*[`pkg/kv/kvserver/replica_test.go` line 8544 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/83213#-N80JqBPAGvBvTkfav1c:-N80JqBPAGvBvTkfav1d:bx47xe9) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb739440c59ef9b7ca739e1f36f7afe104e99a67/pkg/kv/kvserver/replica_test.go#L8544)):*
> ```Go
> 	testutils.SucceedsSoon(t, func() error {
> 		statsAfter = repl.GetMVCCStats()
> 		if statsAfter.ValCount != numVersions {
> ```

This test is validating that updates to the `ValCount` field commutes, but not that updates to every other field commutes properly. I think it would be valuable to have the test run periodic consistency checks as well, other something that recomputes the stats on the range and verifies that they are correct at every step along the way.
___
*[`pkg/kv/kvserver/replica_test.go` line 8828 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/83213#-N80JSjV3H01h4c2iIiV:-N80JSjV3H01h4c2iIiW:btdcblw) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb739440c59ef9b7ca739e1f36f7afe104e99a67/pkg/kv/kvserver/replica_test.go#L8828)):*
> ```Go
> 
> 			// Since the GC request does not acquire latches on the keys being GC'ed,
> 			// they're not guaranteed to wait for these above Puts to get applied. So
> ```

"applied on the leaseholder"

Also, maybe point at `AckCommittedEntriesBeforeApplication`.


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
> You're saying that we need to serialize new MVCC range tombstone writes at a timestamp ts2 and the GC of old MVCC point keys at a timestamp ts1, where ts1 < ts2? Or is this only the case for the GC of old MVCC ranged keys?

The primary concern has been the latter: GC of old MVCC range keys must serialize with overlapping writes of MVCC ranged keys, since the stats updates are not commutative.

I don't believe there should be a need for point GC to serialize with newer MVCC range key writes, since the point key stats are only affected by the point/range key directly above them, and that point/range key must be below the GC threshold for the point key to be eligible for GC in the first place (assuming a non-tombstone point key).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83213#-:-N80XKcjAk-kHYOmENAY:b-xqn4q6)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15, @aliher1911, and @arulajmani)



<!-- Sent from Reviewable.io -->

aayushshah15 reviewed (commented):
> Does this PR currently address this case?

Based on this latest comment by Erik, it does. We're serializing range key GC with overlapping MVCC range key writes (but also unfortunately, overlapping point key reads and writes). @erikgrinaker or @aliher1911, would either of you mind taking a quick look (and stamping) if this all seems copacetic?

 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83213#-:-N8_N0dx6qvmHL3_81jV:b7gqw2a)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aliher1911, @arulajmani, and @nvanbenschoten)

___
*[`pkg/kv/kvserver/batcheval/cmd_gc.go` line 39 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/83213#-N80HT29E9dlzZ2GZqI3:-N8_uV-2-kQkEEAXfU2s:b-dddzu0) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb739440c59ef9b7ca739e1f36f7afe104e99a67/pkg/kv/kvserver/batcheval/cmd_gc.go#L39)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This is new as of https://github.com/cockroachdb/cockroach/commit/285aead75f77740df357a370aff37e91e579ff8a. Do we still need these latches or does the same reasoning that we're using to remove point latches apply to range keys as well? Maybe we should include @aliher1911 as a reviewer here.

EDIT: I see that Erik has discussed this in the thread above. We should work in some of that into the comment here.
</blockquote></details>

Done. Let me know if you think the comment is lacking. 
___
*[`pkg/kv/kvserver/spanset/spanset.go` line 114 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/83213#-N80INw_CySP8zxzh9yu:-N80P5rl0eL8XBjRi6Yt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb739440c59ef9b7ca739e1f36f7afe104e99a67/pkg/kv/kvserver/spanset/spanset.go#L114)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do we need to reset `allowUndeclared` here?
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/spanset/spanset.go` line 403 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/83213#-N80IjVJ9cQmkruwObK9:-N80SAor89X3HxgNWF2y:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb739440c59ef9b7ca739e1f36f7afe104e99a67/pkg/kv/kvserver/spanset/spanset.go#L403)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Can we use this to get rid of `DisableReaderAssertions`?
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/replica_test.go` line 8544 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/83213#-N80JqBPAGvBvTkfav1c:-N80TAXp2KZ4XU5PLUv3:b-dgh84h) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb739440c59ef9b7ca739e1f36f7afe104e99a67/pkg/kv/kvserver/replica_test.go#L8544)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This test is validating that updates to the `ValCount` field commutes, but not that updates to every other field commutes properly. I think it would be valuable to have the test run periodic consistency checks as well, other something that recomputes the stats on the range and verifies that they are correct at every step along the way.
</blockquote></details>

Discussed offline. I've changed the test to concurrently recompute MVCC stats (and ensure that they are never inconsistent at any point) while the GC and writes are running. 
___
*[`pkg/kv/kvserver/replica_test.go` line 8828 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/83213#-N80JSjV3H01h4c2iIiV:-N80SU6vCETgLaqKfJZP:bs7wumx) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb739440c59ef9b7ca739e1f36f7afe104e99a67/pkg/kv/kvserver/replica_test.go#L8828)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

"applied on the leaseholder"

Also, maybe point at `AckCommittedEntriesBeforeApplication`.
</blockquote></details>

Done. 


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
> I don't believe there should be a need for point GC to serialize with newer MVCC range key writes, since the point key stats are only affected by the point/range key directly above them, and that point/range key must be below the GC threshold for the point key to be eligible for GC in the first place (assuming a non-tombstone point key).

This last part about tombstone point keys is interesting. If we have a tombstone and a non-tombstone below the GC threshold, we are able to GC both of them, leaving no remaining version on the key. Based on a reading of `ComputeStatsForRangeWithVisitors` and `MVCCGarbageCollect`, I think this is ok, as the presence of a range tombstone does not have an effect on point tombstones directly below it, only point non-tombstones.

> We could either build one of these, or we could possibly hack around it by taking out a read/write latch at timestamp hlc.MaxTimestamp which would not serialize with any readers (since there cannot be any readers above that timestamp).

Taking out a read/write latch at timestamp `hlc.MaxTimestamp` seems like a reasonable solution to me.

> Reviewed 5 of 6 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83213#-:-N8eeqg80kgbUvkf2KsT:b-tdji7o)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15, @aliher1911, and @arulajmani)

___
*[`pkg/kv/kvserver/replica_test.go` line 8814 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/83213#-N8ei_ETEp-Gx0KPfiNZ:-N8ei_EU3375uWA3kLKd:bhxxyd3) ([raw file](https://github.com/cockroachdb/cockroach/blob/483e83b38c04b3378c4973327ca7959d1ed5dda8/pkg/kv/kvserver/replica_test.go#L8814)):*
> ```Go
> 			// Since the GC request does not acquire latches on the keys being GC'ed,
> 			// they're not guaranteed to wait for these above Puts to get applied on
> 			// the leaseholder. See AckCommittedEntriesBeforeApplication() the comment
> ```

"and the"


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (approved):
This all seems fine to me, thanks for improving it!

@aliher1911 This mostly addresses #84576. I think we should extend some of these benchmarks to include range tombstone GC, and see if we need to either break up or otherwise limit the work we do per request to make sure we're not holding onto the latches for too long. We also need to look at TBI performance below the GC threshold -- at that point, most data is likely compacted down to the lower LSM levels, and so we'll have to rely on skipping blocks rather than entire SSTs, and I don't know if we get as much of a boost from that.

> I think this is ok, as the presence of a range tombstone does not have an effect on point tombstones directly below it, only point non-tombstones.

Yeah, that sounds right.

>Taking out a read/write latch at timestamp hlc.MaxTimestamp seems like a reasonable solution to me.

Thanks for confirming -- I think that's the only viable solution with only a week left before stability. Not too thrilled about the wide write latches though, we'll just have to make sure they're short-lived and maybe break them up. Are they essentially going to form a barrier in the concurrency manager queue, or is there some way we can avoid head-of-line blockage? This reminds me about the problems with split latches (#81152).

> Reviewed 4 of 6 files at r3, 2 of 2 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83213#-:-N8hUYAC0-CYoZGGdybv:bjmdzjb)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15, @aliher1911, and @arulajmani)

___
*[`-- commits` line 59 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/83213#-N8hZREqAMXz80U79TWV):*
This is lovely. 🚀
___
*[`pkg/kv/kvserver/batcheval/cmd_add_sstable.go` line 63 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/83213#-N8hXIfECFcrl2GNvGEf:-N8hXIfF6bHpFFgm0gqj:btg5uwx) ([raw file](https://github.com/cockroachdb/cockroach/blob/af7e8d6acc0660acf4531ff079c17c285c269854/pkg/kv/kvserver/batcheval/cmd_add_sstable.go#L63)):*
> ```Go
> 		latchSpans.AddMVCC(spanset.SpanReadOnly, roachpb.Span{Key: l, EndKey: r}, header.Timestamp)
> 	}
> 	latchSpans.DisableUndeclaredAccessAssertions()
> ```

Any particular reason for this? I'd rather disable the assertions for the one specific access that violates the latches, than the entire request.
___
*[`pkg/kv/kvserver/batcheval/cmd_recompute_stats.go` line 57 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/83213#-N8hYjowD_7FHg1gqkf9:-N8hYjowD_7FHg1gqkfA:bdgxxsi) ([raw file](https://github.com/cockroachdb/cockroach/blob/af7e8d6acc0660acf4531ff079c17c285c269854/pkg/kv/kvserver/batcheval/cmd_recompute_stats.go#L57)):*
> ```Go
> 	latchSpans.AddNonMVCC(spanset.SpanReadWrite, roachpb.Span{Key: keys.TransactionKey(rdKey, uuid.Nil)})
> 	// Disable the assertions which check that all reads were previously declared.
> 	latchSpans.DisableUndeclaredAccessAssertions()
> ```

As an apropos, I have been a bit worried about the non-commutativity of MVCC range tombstones in the context of consistency check stats calculations. As far as I can tell, this should still be safe, because the stats computation is only reading (from a stable Pebble snapshot) and the stats delta of the write will be correct regardless of the base stats. The non-commutativity is only a problem with concurrent writers. Anything I'm missing?


<!-- Sent from Reviewable.io -->

aayushshah15 commented:
*[`pkg/kv/kvserver/batcheval/cmd_add_sstable.go` line 63 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/83213#-N8hXIfECFcrl2GNvGEf:-N8zBR1f5Rti_XiItnPo:b-2l35mk) ([raw file](https://github.com/cockroachdb/cockroach/blob/af7e8d6acc0660acf4531ff079c17c285c269854/pkg/kv/kvserver/batcheval/cmd_add_sstable.go#L63)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Any particular reason for this? I'd rather disable the assertions for the one specific access that violates the latches, than the entire request.
</blockquote></details>

Are you suggesting that we keep `spanset.DisableReaderAssertions` around, or something else? 


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83213)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15, @aliher1911, and @arulajmani)

___
*[`pkg/kv/kvserver/batcheval/cmd_add_sstable.go` line 63 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/83213#-N8hXIfECFcrl2GNvGEf:-N8zCbn-0A86MxFayfMU:bc9j1pi) ([raw file](https://github.com/cockroachdb/cockroach/blob/af7e8d6acc0660acf4531ff079c17c285c269854/pkg/kv/kvserver/batcheval/cmd_add_sstable.go#L63)):*
<details><summary><i>Previously, aayushshah15 (Aayush Shah) wrote…</i></summary><blockquote>

Are you suggesting that we keep `spanset.DisableReaderAssertions` around, or something else? 
</blockquote></details>

Yeah, for the `AddSSTable` case there was only a single read access that needed to circumvent the spanset assertion (and it was safe to do so). For the rest of the accesses, I think we'd rather still keep the assertions enabled, if possible.


<!-- Sent from Reviewable.io -->

aayushshah15 reviewed (commented):
Thanks for the reviews! 

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83213#-:-N8zFBm0AnB7Ea8LnhMw:b-w4qx27)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15, @aliher1911, @arulajmani, @erikgrinaker, and @nvanbenschoten)

___
*[`pkg/kv/kvserver/replica_test.go` line 8814 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/83213#-N8ei_ETEp-Gx0KPfiNZ:-N8zEOydEjb0oRyMAq1s:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/483e83b38c04b3378c4973327ca7959d1ed5dda8/pkg/kv/kvserver/replica_test.go#L8814)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

"and the"
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/batcheval/cmd_add_sstable.go` line 63 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/83213#-N8hXIfECFcrl2GNvGEf:-N8zDuNV1HV7OC5S8bBf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/af7e8d6acc0660acf4531ff079c17c285c269854/pkg/kv/kvserver/batcheval/cmd_add_sstable.go#L63)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Yeah, for the `AddSSTable` case there was only a single read access that needed to circumvent the spanset assertion (and it was safe to do so). For the rest of the accesses, I think we'd rather still keep the assertions enabled, if possible.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

aayushshah15 commented:
bors r+
